### DISCUSSION TITLE
Need for a 'browse directory' widget

### DISCUSSION THREAD

**Discussion by wilsonk** [david-vanderson/dvui#23] [Ideas]
I am just wondering if anyone else finds a need for this? And how difficult it might be to implement. Or is there some widget that can be massaged into acting in this capacity and I am unaware of it?

**Comment by david-vanderson** [comment]
There is not currently a widget for this, but there should be.  The difficulty depends greatly on how much you want it to do.  Traditional "file open" dialogs (gtk, win32, etc.) have a lot of stuff in them and integrate into the rest of the desktop (showing recent files for example).

Still, we should have one.  It doesn't need to do everything at the start, and getting experience doing the basics is very useful.

Can you say more about what you need it to do?  Are you thinking a modal "file open" dialog?  Or some kind of file preview thing?

**Comment by wilsonk** [comment]
Actually, I was just thinking of a simple directory navigation dialog to start. Doesn't even have to pick or open a file really, just be able to pick a directory after navigation (so the '.' and '..' are working, etc.). So, maybe a label or textentry, paired with a 'directory' button (a button a 'directory' icon on it) at the end such, that picking a directory updates the label/textentry? Something along those lines to start. I assume a modal dialog popup.

Simple file picking wouldn't be too much different...but then there are a few more questions for a full featured picker surrounding mime types, recent files, icons, and all that jazz to consider, I suppose.

**Comment by david-vanderson** [comment]
Sounds great!  I think options for where to put it during development:
- sdl-test.zig until it's moved to the demo window
- or we add a "in development" section to the demo window

Let me know if you get the start of something going!

**Comment by VisenDev** [comment]
I was recently thinking about this recently as well. If we want to directly create a file picker - I think the best and simplest approach would be to use a cross platform native file dialog library. This will allow users to open their native file picker.

The advantage of this is a) we don't have to implement our own file picker, and b) users will be more comfortable using their native file picker than something custom. 

There are single header c libraries like https://github.com/samhocevar/portable-file-dialogs that we could use for this purpose

If we go this route, a function like `openFileDialog` should probably be added to the backend interface, so that more unique backends like `web/wasm` can be accommodated

**Comment by david-vanderson** [comment]
That sounds like a great idea to me.  I will do some investigation of how to integrate this.

Currently I'm thinking we add a modal dvui dialog at the same time we launch the native dialog, and do the communication between those two, and then results are pulled from the dvui dialog.  Not sure how the details will work.

**Comment by david-vanderson** [comment]
Filed issue https://github.com/david-vanderson/dvui/issues/93
